### PR TITLE
[feat] experimental.nn.multiprocess_pipe: re-implemented using rpc

### PR DESCRIPTION
**Pull request by msbaines**
Re-implemented nn.MultiProcessPipe based on the rpc pipeline parallel
tutorial:

(https://pytorch.org/tutorials/intermediate/dist_pipeline_parallel_tutorial.html).

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by msbaines** [comment]
@mrshenli @pritamdamania

File: fairscale/experimental/nn/multiprocess_pipe.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+BOUNCE_TENSORS = True
+
+
+def _verify_module(module: List[LayerSpec]) -> None:
+    if not isinstance(module, List):
+        raise TypeError("module must be a list")
+
+    for elem in module:
+        if not isinstance(elem, tuple):
+            raise TypeError("module must be a list of tuple")
```
Comment (anj-s) [inline fairscale/experimental/nn/multiprocess_pipe.py:34]: Should we say "elem must be of ..."?

Comment (anj-s) [reply fairscale/experimental/nn/multiprocess_pipe.py:34]: Also, does it make sense to print out what was passed in as part of the error message? just for easier debugging.

File: fairscale/experimental/nn/multiprocess_pipe.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    chunks: int = 1
+
+    #: The checkpoint mode to determine when to enable checkpointing. It is one
+    #: of ``'always'``, ``'except_last'``, or ``'never'``.
+    checkpoint: str = "never"
+
+    def __init__(
+        self,
+        module: List[LayerSpec],
+        *,
```
Comment (anj-s) [inline fairscale/experimental/nn/multiprocess_pipe.py:148]: Is this right? what does a single * mean?

Comment (msbaines) [reply fairscale/experimental/nn/multiprocess_pipe.py:148]: Everything after the '*' must be passed as a keyword argument.

File: fairscale/experimental/nn/multiprocess_pipe.py (hunk 1 above)
Comment (msbaines) [reply fairscale/experimental/nn/multiprocess_pipe.py:34]: module is correct in this context. module is the name constructor param. The error is telling the API user they passed in a value of the wrong type.

Comment (msbaines) [reply fairscale/experimental/nn/multiprocess_pipe.py:34]: > Also, does it make sense to print out what was passed in as part of the error message? just for easier debugging.

I usually like to print things out but in this case is a constructor parameter so it should fail pretty quickly. If it fails, it essentially means you are using the API incorrectly.

File: fairscale/experimental/nn/multiprocess_pipe.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+
+class _ToHere(Module):
+    def __init__(self, device: str):
+        super().__init__()
+        self.device = device
+
+    def forward(self, x_rref: rpc.RRef) -> Tensor:  # type: ignore
+        if BOUNCE_TENSORS:
+            return x_rref.remote().cpu().to_here().to(self.device)
```
Comment (sidgoyal78) [inline fairscale/experimental/nn/multiprocess_pipe.py:54]: I was hoping to understand this better, this moves the tensor to the cpu first?

File: fairscale/experimental/nn/multiprocess_pipe.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+            workers.append(worker)
+            rmodule.append(rlayer)
+
+        self.chunks = chunks
+        self.checkpoint = checkpoint
+        self.module = module
+        self.workers = workers
+        self.rmodule = rmodule
+
+    def forward(self, x: Tensor) -> rpc.RRef:  # type: ignore
```
Comment (sidgoyal78) [inline fairscale/experimental/nn/multiprocess_pipe.py:187]: If i am understanding correctly, this forward() impl doesn't support checkpointing right?

**Review by sidgoyal78 — COMMENTED** [review]
Thanks for the PR. I have left a few questions.

File: fairscale/experimental/nn/multiprocess_pipe.py (hunk 3 above)
Comment (msbaines) [reply fairscale/experimental/nn/multiprocess_pipe.py:54]: rpc (until v1.8.0) only supported cpu tensors so I had to bounce to cpu first. I recently learned of GPU support in v1.8.0 which I am now trying to incorporate.

File: fairscale/experimental/nn/multiprocess_pipe.py (hunk 4 above)
Comment (msbaines) [reply fairscale/experimental/nn/multiprocess_pipe.py:187]: It does not yet although it may support PyTorch's regular checkpointing. I will test that.

**Review by anj-s — APPROVED** [review]
Thank you for this awesome change! It is so simple to comprehend. Just curious if you had a chance to see how the perf compares to the current implementation.

File: tests/experimental/nn/test_multiprocess_pipe.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    assert y.shape == torch.Size([8, 4])
+    assert y.requires_grad is True
+    assert torch.allclose(y.sum(), expected_sum), f"{y.sum()} != {expected_sum}"
+
+
+@rpc_test(world_size=2)
+@pytest.mark.parametrize("devices", DEVICES)
+def backward(devices):
+    torch.random.manual_seed(3)
+    criterion = DistributedLoss(torch.nn.MSELoss)
```
Comment (pritamdamania87) [inline tests/experimental/nn/test_multiprocess_pipe.py:153]: `DistributedLoss` is really a nice abstraction to deal with this problem!

**Pull request merged**